src/PowerShellEditorServices/Services/TextDocument/Handlers/PsesSemanticTokens.cs

```diff
@@ -0,0 +1,175 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Management.Automation.Language;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.PowerShell.EditorServices.Services;
+using Microsoft.PowerShell.EditorServices.Services.TextDocument;
+using Microsoft.PowerShell.EditorServices.Utility;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document.Proposals;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models.Proposals;
+
+namespace Microsoft.PowerShell.EditorServices.Handlers
+{
+    internal class PsesSemanticTokens : SemanticTokensHandler
+    {
+        private readonly ILogger _logger;
+        private readonly WorkspaceService _workspaceService;
+        private static readonly SemanticTokensRegistrationOptions s_registrationOptions = new SemanticTokensRegistrationOptions() {
+            DocumentSelector = LspUtils.PowerShellDocumentSelector,
+            Legend = new SemanticTokensLegend(),
+            DocumentProvider = new Supports<SemanticTokensDocumentProviderOptions>(isSupported: true,
+                new SemanticTokensDocumentProviderOptions {
+                    Edits = true
+                }),
+            RangeProvider = true
+        };
+
+        public PsesSemanticTokens(ILogger<PsesSemanticTokens> logger, WorkspaceService workspaceService) : base(s_registrationOptions)
+        {
+            _logger = logger;
+            _workspaceService = workspaceService;
+        }
+
+        protected override Task Tokenize(SemanticTokensBuilder builder, ITextDocumentIdentifierParams identifier,
+            CancellationToken cancellationToken)
+        {
+            ScriptFile file = _workspaceService.GetFile(DocumentUri.GetFileSystemPath(identifier));
+            Token[] tokens = file.ScriptTokens;
+            foreach (var token in tokens){
+                PushToken(token, builder);
+            }
+            return Task.CompletedTask;
+        }
+
+        private static void PushToken(Token token, SemanticTokensBuilder builder)
+        {
+
+            List<SemanticToken> semanticToken = ConvertToSemanticTokens(token);
+            foreach(SemanticToken sToken in semanticToken)
+            {
+                builder.Push(sToken.Line, sToken.Index, length: sToken.Text.Length,
+                    sToken.Type, tokenModifiers: sToken.TokenModifiers);
+            }
```

Comment:
The `ConvertToSemanticTokens()` method will allocate a new list for each token, only to push those into the builder immediately.

If we eliminate the intermediate datastructure, we can prevent a number of allocations and therefore help PSES spend less time/CPU on garbage collection.

Two suggestions on how to do this are:

1. Inline the `ConvertToSemanticTokens()` method into `PushToken()`. This gives you more power to set simple and efficient semantics for token creation, but is less composable and harder to test
2. Turn `ConvertToSemanticTokens()` into an [iterator method](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/iterators). This depends on slightly more complex semantics and in older .NET Framework versions is less efficient, but allows you to largely keep the code structure you have now and the composability it buys you, and still buys some extra efficiency beyond creating a new `List<>` each time (especially usually a singleton or empty list)